services/brig/src/Brig/User/Auth.hs

```diff
@@ -252,12 +254,32 @@ isPendingActivation ident = case ident of
             Just SSOIdentity {} -> False -- sso-created users are activated immediately.
             Nothing -> True
 
+-- | Validate a list of (User/LH) tokens potentially with an associated access token.
+--   If there are multiple valid cookies, we try all of them. When an access token is
+--   given, we perform the usual checks.
+--   If multiple cookies are given and several are valid, we return the first valid one.
 validateTokens ::
+  ZAuth.TokenPair u a =>
+  List1 (ZAuth.Token u) ->
+  Maybe (ZAuth.Token a) ->
+  ExceptT ZAuth.Failure AppIO (UserId, Cookie (ZAuth.Token u))
+validateTokens uts at = do
+  tokens <- forM uts $ \ut -> lift $ runExceptT (validateToken ut at)
+  getFirstSuccessOrFirstFail tokens
+  where
+    -- FUTUREWORK: There is surely a better way to do this
+    getFirstSuccessOrFirstFail :: List1 (Either ZAuth.Failure (UserId, Cookie (ZAuth.Token u))) -> ExceptT ZAuth.Failure AppIO (UserId, Cookie (ZAuth.Token u))
```

Comment:
This function is fully pure now; and can be way more polymorphic.  This also very much signifies the future work.

The callsite above needs to just be replaced with:
```haskell
ExceptT $ getFirstSuccessOrFirstFail tokens
```

```suggestion
    getFirstSuccessOrFirstFail :: List1 (Either ZAuth.Failure (UserId, Cookie (ZAuth.Token u))) -> ExceptT ZAuth.Failure AppIO (UserId, Cookie (ZAuth.Token u))
    getFirstSuccessOrFirstFail :: List1 (Either e a) -> Either e a
```